src/HexagonOptimize.cpp

```diff
@@ -1605,6 +1606,271 @@ class OptimizeShuffles : public IRMutator {
 public:
     OptimizeShuffles(int lut_alignment) : lut_alignment(lut_alignment) {}
 };
+
+class VtmpyGenerator : public IRMutator {
+private:
+    using IRMutator::visit;
+    Scope<Expr> vars;
+    typedef pair<Expr, size_t> LoadIndex;
+
+    // Return value of variable if expr is a variable.
+    // Else return the expr.
+    Expr get_val(Expr e) {
+        const Variable *op = e.as<Variable>();
+        if (op) {
+            if (vars.contains(op->name)) {
+                return vars.get(op->name);
+            }
+            else {
+                return Expr();
+            }
+        } else {
+            return e;
+        }
+    }
+
+    // Save variable value
+    void visit(const Let* op) {
+        vars.push(op->name, op->value);
+        IRMutator::visit(op);
+        vars.pop(op->name);
+    }
+
+    // Save variable value
+    void visit(const LetStmt *op) {
+        vars.push(op->name, op->value);
+        IRMutator::visit(op);
+        vars.pop(op->name);
+    }
+
+    // Check if vectors a and b point to the same buffer with the base of a
+    // shifted by diff i.e. base(a) = base(b) + diff.
+    bool is_base_shifted(const Expr &a, const Expr &b, int diff) {
+        Expr maybe_load_a = calc_load(a);
+        Expr maybe_load_b = calc_load(b);
+
+        if (maybe_load_a.defined() && maybe_load_b.defined()) {
+            const Load* load_a = maybe_load_a.as<Load>();
+            const Load* load_b = maybe_load_b.as<Load>();
+            const Ramp* ramp_a = get_val(load_a->index).as<Ramp>();
+            const Ramp* ramp_b = get_val(load_b->index).as<Ramp>();
+            if (ramp_a && ramp_b && load_a->name == load_b->name) {
+                Expr base_diff = simplify(ramp_a->base - ramp_b->base - diff);
+                if (is_const(base_diff, 0)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    // Return the load expression of first vector if all vector in exprs are
+    // contiguous vectors pointing to the same buffer.
+    Expr are_contiguous_vectors(const vector<Expr> exprs) {
+        if (exprs.size() == 0) {
+            return Expr();
+        }
+        // If the shuffle simplifies then the vectors are contiguous.
```

Comment:
I was thinking one way to make this work is to just substitute all lets (https://github.com/halide/Halide/blob/master/src/Substitute.h#L55), use the Shuffle test, and then re-run CSE after. Other passes do something similar... @abadams any comment on this? I'd really like to avoid the ~100 lines of very logic-dense code when the Shuffle test should be able to handle it.